nengo/spa/state.py

```diff
@@ -0,0 +1,76 @@
+import numpy as np
+
+import nengo
+from nengo.spa.module import Module
+from nengo.networks.ensemblearray import EnsembleArray
+
+# Is the feedback parameter a big mistake or should I change the tests?
+
+class State(Module):
+    """A module capable of representing a single vector, with optional memory.
+
+    This is a minimal SPA module, useful for passing data along (for example,
+    visual input).
+
+    Parameters
+    ----------
+    dimensions : int
+        Number of dimensions for the vector
+    subdimensions : int
+        Size of the individual ensembles making up the vector.  Must divide
+        evenly into dimensions
+    neurons_per_dimensions : int
+        Number of neurons in an ensemble will be this*subdimensions
+    vocab : Vocabulary, optional
+        The vocabulary to use to interpret this vector
+    direct : boolean
+        Whether or not to use direct mode for the neurons
+    """
+
+    def __init__(self, dimensions, subdimensions=16, neurons_per_dimension=50,
+                 feedback=1.0, tau=0.1,
```

Comment:
Hmm...  I don't think a parameter called ```feedback``` should also change the input transform, as you have it here.  I'd expect a parameter called ```feedback``` to only affect the feedback.

In the previous code, the two parameters were ```synapse``` and ```tau```, where ```synapse``` was the "synaptic filter to use on recurrent connection" and ```tau``` was the "effective time constant of the integrator".  In other words, if ```tau=3.0``` then the memory would last for around 3 seconds (and if ```tau=None``` then it would try to be a perfect integrator with an effective time constant of infinity).  That's what all the scaling was about was attempting to make that happen.

Part of me really likes that way of organizing it.  But part of me also thinks that it was a bit overcomplicated, difficult to explain, and never really worked that well anyway (as in the actual time constant you get isn't all that close to what you asked for, and it's hard to know over what range of values it will be good).  And in practice, we always ended up doing something like "just set the recurrent transform to 0.95".

So all of that said, I think the simplest and clearest approach is to do a ```feedback``` parameter.  I think it should default to ```0.0``` (i.e. no feedback) and if people want a memory they should set it to ```1.0```.  And if hey want a decaying memory they can set it to ```0.95``` or whatever.  And the only thing that parameter should affect is that feedback (i.e. it should not affect the input scaling).

The other option which is a bit more complex is to just have the ```tau``` parameter, or perhaps ```memory_time``` and do the same ```transform = 1.0 - synapse/memory_time``` calculation (although that in itself only works for ```synapse``` values that are floats....)

So yeah, I'd lean more to the simple option.  The tau thing was a neat idea, but not worth in the long run, I think.